### PR TITLE
Rebased the feature/metrics branch from master manually

### PR DESCRIPTION
**Pull request by purbon**
This has been done  into this new branch. This branch include code for the metrics delivery,
instrumentalization and the web api to be presented to the user. It also
include test for each component implemented.

@ph can you make sure everything is reasonable?

### CODE REVIEW

**Comment by purbon** [comment]
@ph, saw this error

```
 1) LogStash::Runner argument parsing with no arguments should show help
     Failure/Error: subject.run(args)
       (#<Cabin::Channel:0x76206c9a>).warn("You may be interested in the '--configtest' flag which you can use to validate\nlogstash's configuration before you choose to restart a running system.")
           expected: 1 time with any arguments
           received: 2 times with arguments: ("You may be interested in the '--configtest' flag which you can use to validate\nlogstash's configuration before you choose to restart a running system.")
     # ./logstash-core/lib/logstash/runner.rb:165:in `execute'
     # ./vendor/bundle/jruby/1.9/gems/clamp-0.6.5/lib/clamp/command.rb:67:in `run'
     # ./logstash-core/spec/logstash/runner_spec.rb:40:in `(root)'
     # ./vendor/bundle/jruby/1.9/gems/rspec-wait-0.0.8/lib/rspec/wait.rb:46:in `(root)'
     # ./rakelib/test.rake:42:in `(root)'
     # ./vendor/bundle/jruby/1.9/gems/rake-10.4.2/lib/rake/task.rb:240:in `execute'
     # ./vendor/bundle/jruby/1.9/gems/rake-10.4.2/lib/rake/task.rb:235:in `execute'
     # ./vendor/bundle/jruby/1.9/gems/rake-10.4.2/lib/rake/task.rb:179:in `invoke_with_call_chain'
     # ./vendor/bundle/jruby/1.9/gems/rake-10.4.2/lib/rake/task.rb:172:in `invoke_with_call_chain'
     # ./vendor/bundle/jruby/1.9/gems/rake-10.4.2/lib/rake/task.rb:165:in `invoke'
     # ./vendor/bundle/jruby/1.9/gems/rake-10.4.2/lib/rake/application.rb:150:in `invoke_task'
     # ./vendor/bundle/jruby/1.9/gems/rake-10.4.2/lib/rake/application.rb:106:in `top_level'
     # ./vendor/bundle/jruby/1.9/gems/rake-10.4.2/lib/rake/application.rb:106:in `top_level'
     # ./vendor/bundle/jruby/1.9/gems/rake-10.4.2/lib/rake/application.rb:115:in `run_with_threads'
     # ./vendor/bundle/jruby/1.9/gems/rake-10.4.2/lib/rake/application.rb:100:in `top_level
```

before, but some times it was gone .... do you understand where might be coming from?

**Comment by purbon** [comment]
please jenkins, test this.

**Comment by purbon** [comment]
you see, now good .... i see a possible random effect here we should iron somehow.

**Comment by ph** [comment]
This is the seed for the failing test, (run on jenkins)
Randomized with seed 10771

**Comment by purbon** [comment]
@ph this seed also fails for feature/metrics branch, so looks like is not introduced by this pr.

**Comment by purbon** [comment]
Executing ```skywalker% bin/rspec --seed 10771 logstash-core/spec/logstash/runner_spec.rb``` causes the same error.

**Comment by purbon** [comment]
@ph fixed in last commit. can you review and we merge?

File: logstash-core/lib/logstash/agent.rb [hunk 1]
```diff
@@ -26,26 +39,112 @@ def execute
...
+
+    metric_pipeline_config =<<-EOS
+      input {
+        metrics {}
+      }
+      output {
+        elasticsearch {
+          flush_size => 1
+          hosts => "127.0.0.1"
+          index => "metrics-%{YYYY.MM.dd}"
```
Comment (ph) [inline logstash-core/lib/logstash/agent.rb]: `index => "metrics-%{YYYY.MM.dd}"` should be     `index => "metrics-%{+YYYY.MM.dd}"`

**Comment by ph** [comment]
Small typo other than that LGTM.

I have found a major slow down in this revision of the code investigating :)

**Comment by ph** [comment]
The slowdown is only when the internal pipeline is used, might something we can tweak easily :heart:

**Comment by purbon** [comment]
not sure how this could have introduced an slow down? have you check both branches for the same behaviour? the issue at https://github.com/elastic/logstash/pull/4628#issuecomment-180055741 was also in the ```feature/metrics``` branch.

**Comment by purbon** [comment]
please jenkins, test this.

**Pull request closed without merging**